app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java

```diff
@@ -1,114 +1,110 @@
+        Logger.debug("Stopping toast notification listener.");
+        uiAutomation.setOnAccessibilityEventListener(originalListener);
     }
 
-    private class Listener extends Thread {
-
-        //return true if the AccessibilityEvent type is NOTIFICATION type
-        private final UiAutomation.AccessibilityEventFilter eventFilter = new UiAutomation.AccessibilityEventFilter() {
-            @Override
-            public boolean accept(AccessibilityEvent event) {
-                return event.getEventType() == AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED;
-            }
-        };
-        private final Runnable runnable = new Runnable() {
-            @Override
-            public void run() {
-                // Not performing any event.
-            }
-        };
-        private boolean stopLooping = false;
-        private long previousTime = currentTimeMillis();
-
-        public void run() {
-            while (true) {
-                AccessibilityEvent accessibilityEvent = null;
-                toastMessages = init();
-
-                try {
-                    //wait for AccessibilityEvent filter
-                    accessibilityEvent = UiAutomatorBridge.getInstance().getUiAutomation()
-                            .executeAndWaitForEvent(runnable /*executable event*/, eventFilter /* event to filter*/, WAIT_FOR_EVENT_TIMEOUT /*time out in ms*/);
-                } catch (Exception ignore) {
-                }
-
-                if (accessibilityEvent != null) {
-                    toastMessages = accessibilityEvent.getText();
-                    previousTime = currentTimeMillis();
-                }
-                if (stopLooping) {
-                    break;
-                }
+    @Override
+    public synchronized void onAccessibilityEvent(AccessibilityEvent event) {
+        if (event.getEventType() == AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED) {
+            Logger.debug("Catch toast message: " + event);
+            List<CharSequence> text = event.getText();
+            if (text != null && !text.isEmpty()) {
+                setToastMessage(event.getText());
             }
         }
 
-        public List<CharSequence> init() {
-            if (currentTimeMillis() - previousTime > TOAST_CLEAR_TIMEOUT) {
-                return new ArrayList<CharSequence>();
-            }
-            return toastMessages;
+        if (originalListener != null) {
+            originalListener.onAccessibilityEvent(event);
         }
+    }
 
-        public void stopLooping() {
-            stopLooping = true;
+    public boolean isListening() {
+        return uiAutomation.getOnAccessibilityEventListener() == this;
+    }
+
+    protected long getToastClearTimeout() {
+        return TOAST_CLEAR_TIMEOUT;
+    }
+
+    @NonNull
+    protected List<CharSequence> getToastMessage() {
+        if (!toastMessage.isEmpty() && currentTimeMillis() - recentToastTimestamp > getToastClearTimeout()) {
```

Comment:
is it really necessary to delete the message after the timeout? I'd rather collect toasts together with their timestamps and let the client to decide whether he needs them or not by analyzing the timestamp

Reply:
I'll be happy if someone implements this behaviour :)

Reply:
I believe you can easily handle it on your own ;)